File: server/src/main/java/com/securechat/server/RoomManager.java
package com.securechat.server;

import com.securechat.common.dto.MessageDTO;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.*;
import java.util.concurrent.ConcurrentHashMap;

/**
 * RoomManager – Quản lý phòng chat và routing tin nhắn.
 *
 * - Lưu danh sách client đang kết nối (userId → ClientHandler)
 * - Lưu danh sách thành viên mỗi phòng (roomId → Set<userId>)
 * - Broadcast tin nhắn đến đúng người nhận
 */
public class RoomManager {

    private static final Logger log = LoggerFactory.getLogger(RoomManager.class);

    /** userId → ClientHandler (tất cả client đang online) */
    private final Map<Integer, ClientHandler> onlineClients =
            new ConcurrentHashMap<>();

    /** roomId → Set<userId> (thành viên mỗi phòng) */
    private final Map<Integer, Set<Integer>> roomMembers =
            new ConcurrentHashMap<>();

    // ────────────────────────────────────────────────────────────
    // Quản lý client
    // ────────────────────────────────────────────────────────────

    public void registerClient(int userId, ClientHandler handler) {
        onlineClients.put(userId, handler);
        log.info("Client đăng ký: userId={}, tổng online={}",
                userId, onlineClients.size());
    }

    public void removeClient(int userId) {
        onlineClients.remove(userId);
        // Xoá khỏi tất cả các phòng
        roomMembers.values().forEach(members -> members.remove(userId));
        log.info("Client rời: userId={}, tổng online={}",
                userId, onlineClients.size());
    }

    public boolean isOnline(int userId) {
        return onlineClients.containsKey(userId);
    }

    public Collection<ClientHandler> getOnlineClients() {
        return onlineClients.values();
    }

    // ────────────────────────────────────────────────────────────
    // Quản lý phòng
    // ────────────────────────────────────────────────────────────

    public void joinRoom(int roomId, int userId, ClientHandler handler) {
        roomMembers.computeIfAbsent(roomId, k -> ConcurrentHashMap.newKeySet())
                   .add(userId);
        onlineClients.put(userId, handler);
    }

    public void leaveRoom(int roomId, int userId) {
        Set<Integer> members = roomMembers.get(roomId);
        if (members != null) {
            members.remove(userId);
        }
    }

    public Set<Integer> getRoomMembers(int roomId) {
        return roomMembers.getOrDefault(roomId, Collections.emptySet());
    }

    /**
     * Tạo hoặc lấy phòng chat riêng tư giữa 2 người.
     * RoomId = min(a,b)*100000 + max(a,b) (đảm bảo duy nhất)
     */
    public int getOrCreatePrivateRoom(int userId1, int userId2) {
        int minId = Math.min(userId1, userId2);
        int maxId = Math.max(userId1, userId2);
        int virtualRoomId = minId * 100_000 + maxId;

        roomMembers.computeIfAbsent(virtualRoomId, k -> {
            Set<Integer> members = ConcurrentHashMap.newKeySet();
            members.add(userId1);
            members.add(userId2);
            return members;
        });

        return virtualRoomId;
    }

    // ────────────────────────────────────────────────────────────
    // Broadcast
    // ────────────────────────────────────────────────────────────

    /**
     * Gửi tin nhắn đến tất cả thành viên trong phòng (trừ người gửi).
     */
    public void broadcastToRoom(int roomId, MessageDTO msg, int excludeUserId) {
        Set<Integer> members = getRoomMembers(roomId);
        if (members.isEmpty()) {
            // Nếu không có ai trong phòng, gửi trực tiếp đến receiver
            // (với private chat roomId đã encode cả hai userId)
            broadcastToAll(msg, excludeUserId);
            return;
        }

        int count = 0;
        for (int memberId : members) {
            if (memberId == excludeUserId) continue;
            ClientHandler handler = onlineClients.get(memberId);
            if (handler != null) {
                handler.sendMessage(msg);
                count++;
            }
        }
        log.debug("Broadcast roomId={} → {} clients", roomId, count);
    }

    /**
     * Gửi tin nhắn đến tất cả client đang online (trừ người gửi).
     * Dùng cho SYSTEM messages (user joined/left).
     */
    public void broadcastToAll(MessageDTO msg, int excludeUserId) {
        int count = 0;
        for (Map.Entry<Integer, ClientHandler> entry : onlineClients.entrySet()) {
            if (entry.getKey() == excludeUserId) continue;
            entry.getValue().sendMessage(msg);
            count++;
        }
        log.debug("Broadcast ALL → {} clients", count);
    }

    /**
     * Gửi tin nhắn trực tiếp đến một user cụ thể.
     */
    public boolean sendToUser(int targetUserId, MessageDTO msg) {
        ClientHandler handler = onlineClients.get(targetUserId);
        if (handler != null) {
            handler.sendMessage(msg);
            return true;
        }
        return false; // User offline
    }
}

File: SecureChat.Server/RoomManager.cs
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using SecureChat.Common.DTO;
using SecureChat.Common.Crypto;

namespace SecureChat.Server
{
    public class RoomManager
    {
        // Map: userId -> ClientHandler (Tất cả user đang online)
        private readonly ConcurrentDictionary<int, ClientHandler> _onlineClients = new();

        // Map: roomId -> Set of userIds (Sử dụng ConcurrentDictionary<int, byte> làm Set)
        private readonly ConcurrentDictionary<int, ConcurrentDictionary<int, byte>> _roomMembers = new();

        public void RegisterClient(int userId, ClientHandler handler)
        {
            _onlineClients[userId] = handler;
            Console.WriteLine($"[RoomManager] Đăng ký client: userId={userId}, tổng online={_onlineClients.Count}");
        }

        public void RemoveClient(int userId)
        {
            _onlineClients.TryRemove(userId, out _);
            foreach (var members in _roomMembers.Values)
            {
                members.TryRemove(userId, out _);
            }
            Console.WriteLine($"[RoomManager] Hủy đăng ký client: userId={userId}, tổng online={_onlineClients.Count}");
        }

        public bool IsOnline(int userId)
        {
            return _onlineClients.ContainsKey(userId);
        }

        public IEnumerable<ClientHandler> GetOnlineClients()
        {
            return _onlineClients.Values;
        }

        public void JoinRoom(int roomId, int userId, ClientHandler handler)
        {
            var members = _roomMembers.GetOrAdd(roomId, _ => new ConcurrentDictionary<int, byte>());
            members.TryAdd(userId, 0);
            _onlineClients[userId] = handler;
        }

        public void LeaveRoom(int roomId, int userId)
        {
            if (_roomMembers.TryGetValue(roomId, out var members))
            {
                members.TryRemove(userId, out _);
            }
        }

        public HashSet<int> GetRoomMembers(int roomId)
        {
            if (_roomMembers.TryGetValue(roomId, out var members))
            {
                return new HashSet<int>(members.Keys);
            }
            return new HashSet<int>();
        }

        /// <summary>
        /// Tạo hoặc lấy phòng chat riêng tư ảo giữa 2 người dùng.
        /// RoomId = min(a,b)*100000 + max(a,b) (Đảm bảo ID phòng là duy nhất)
        /// </summary>
        public int GetOrCreatePrivateRoom(int userId1, int userId2)
        {
            int minId = Math.Min(userId1, userId2);
            int maxId = Math.Max(userId1, userId2);
            int virtualRoomId = minId * 100_000 + maxId;

            // Tạo phòng ảo trong CSDL MySQL nếu chưa tồn tại
            DAO.MessageDAO.CreateRoomIfNotExist(virtualRoomId, $"Private_{minId}_{maxId}");

            _roomMembers.GetOrAdd(virtualRoomId, _ =>
            {
                var members = new ConcurrentDictionary<int, byte>();
                members.TryAdd(userId1, 0);
                members.TryAdd(userId2, 0);
                return members;
            });

            return virtualRoomId;
        }

        /// <summary>
        /// Gửi tin nhắn tới mọi thành viên trong phòng (trừ người gửi).
        /// Thực hiện dịch khóa đối xứng (giải mã bằng khóa người gửi, mã hóa bằng khóa từng người nhận).
        /// </summary>
        public void BroadcastToRoom(int roomId, MessageDTO msg, int excludeUserId)
        {
            var members = GetRoomMembers(roomId);
            if (members.Count == 0)
            {
                // Nếu phòng trống (chưa đăng ký member cụ thể), broadcast tới toàn server
                BroadcastToAll(msg, excludeUserId);
                return;
            }

            // Lấy ClientHandler của người gửi để lấy khóa giải mã
            _onlineClients.TryGetValue(excludeUserId, out var senderHandler);
            byte[]? senderKey = senderHandler?.AesKey;

            string? plaintext = null;
            byte[]? fileBytes = null;

            if (senderKey != null)
            {
                try
                {
                    if (msg.Type == MessageDTO.MessageType.TEXT || msg.Type == MessageDTO.MessageType.VIDEO_FRAME)
                    {
                        plaintext = AESUtil.Decrypt(msg.EncryptedContent, senderKey);
                    }
                    else if (msg.Type == MessageDTO.MessageType.FILE)
                    {
                        byte[] encryptedFileBytes = Convert.FromBase64String(msg.EncryptedContent);
                        fileBytes = AESUtil.DecryptBytes(encryptedFileBytes, senderKey);
                    }
                }
                catch (Exception ex)
                {
                    Console.WriteLine($"[RoomManager] Lỗi dịch khóa (giải mã) tin nhắn từ user={excludeUserId} trong phòng={roomId}: {ex.Message}");
                }
            }

            int count = 0;
            foreach (int memberId in members)
            {
                if (memberId == excludeUserId) continue;
                if (_onlineClients.TryGetValue(memberId, out var handler))
                {
                    byte[]? recipientKey = handler.AesKey;
                    if (recipientKey != null)
                    {
                        try
                        {
                            var msgCopy = new MessageDTO(msg.Type)
                            {
                                RoomId = msg.RoomId,
                                SenderId = msg.SenderId,
                                SenderUsername = msg.SenderUsername,
                                Timestamp = msg.Timestamp,
                                FileName = msg.FileName,
                                FileSize = msg.FileSize
                            };

                            if ((msg.Type == MessageDTO.MessageType.TEXT || msg.Type == MessageDTO.MessageType.VIDEO_FRAME) && plaintext != null)
                            {
                                msgCopy.EncryptedContent = AESUtil.Encrypt(plaintext, recipientKey);
                            }
                            else if (msg.Type == MessageDTO.MessageType.FILE && fileBytes != null)
                            {
                                byte[] encryptedForRecipient = AESUtil.EncryptBytes(fileBytes, recipientKey);
                                msgCopy.EncryptedContent = Convert.ToBase64String(encryptedForRecipient);
                            }
                            else
                            {
                                msgCopy.EncryptedContent = msg.EncryptedContent;
                                msgCopy.PlainContent = msg.PlainContent;
                            }

                            handler.SendMessage(msgCopy);
                            count++;
                        }
                        catch (Exception ex)
                        {
                            Console.WriteLine($"[RoomManager] Lỗi dịch khóa (mã hóa) tin nhắn cho user={memberId}: {ex.Message}");
                        }
                    }
                    else
                    {
                        // Fallback nếu người nhận chưa hoàn tất handshake khóa phiên
                        handler.SendMessage(msg);
                        count++;
                    }
                }
            }
        }

        /// <summary>
        /// Gửi tin nhắn tới tất cả client đang online (trừ người gửi).
        /// </summary>
        public void BroadcastToAll(MessageDTO msg, int excludeUserId)
        {
            int count = 0;
            foreach (var kvp in _onlineClients)
            {
                if (kvp.Key == excludeUserId) continue;
                kvp.Value.SendMessage(msg);
                count++;
            }
        }

        /// <summary>
        /// Gửi tin nhắn trực tiếp tới một user cụ thể.
        /// </summary>
        public bool SendToUser(int targetUserId, MessageDTO msg)
        {
            if (_onlineClients.TryGetValue(targetUserId, out var handler))
            {
                handler.SendMessage(msg);
                return true;
            }
            return false;
        }
    }
}
